### PR TITLE
sync with S3

### PR DESCRIPTION
**Pull request by jaymccon**
## Overview

Currently taskcat simply pushes all local files to a bucket on each run, regardless of whether files have changed, it also leaves files in s3 that are no longer in the local project.

This changes behavior to upload only changed files, and delete files not present locally. Speeds up upload when only small changes have been made.

adds `--upload-only` switch that syncs s3 and then exits.

## Testing/Steps taken to ensure quality

launched a quickstart

### Notes

Uses undocumented behavior for checksums of multipart uploads, see https://github.com/aws/aws-cli/issues/2585#issue-226758933

No longer fetching template and parameters from s3, but fetching locally instead. 

enforces template path to `templates/` and parameters to `ci/`, not sure if there are any existing quick starts that don't follow this pattern.

No unit tests yet 😞

Could use a global override option for bucket name, this way a user can use a single bucket for all their taskcat work.

## Testing Instructions

tested against the openshift quick start.

### CODE REVIEW

File: taskcat/s3_sync.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            if relpath == './':
+                relpath = ""
+            for p in S3Sync._exclude_path_prefixes:
+                if relpath.startswith(p):
+                    exclude_path = True
+                    break
+            if not exclude_path:
+                for file in files:
+                    exclude = False
+                    for p in S3Sync._exclude_files:
```
Comment (agamzn) [inline taskcat/s3_sync.py]: Can this be refactored? Looping over the `S3Sync._exclude_files` list for each file seems costly.

File: taskcat/s3_sync.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    def _get_local_file_list(self, path, include_checksums=True):
+        file_list = []
+        path = os.path.abspath(os.path.expanduser(path))
+        for root, dirs, files in os.walk(path):
+            relpath = os.path.relpath(root, path) + "/"
+            exclude_path = False
+            if relpath == './':
+                relpath = ""
+            for p in S3Sync._exclude_path_prefixes:
```
Comment (agamzn) [inline taskcat/s3_sync.py]: Ditto below

File: taskcat/s3_sync.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                resp = self.s3_client.list_objects_v2(Bucket=bucket, Prefix=prefix, ContinuationToken=continuation_token)
+            if 'Contents' in resp:
+                objects += [[i['Key'][len(prefix):], i['ETag']] for i in resp['Contents']]
+            if 'NextContinuationToken' in resp.keys():
+                continuation_token = resp['NextContinuationToken']
+            else:
+                more_objects = False
+        return objects
+
+    def _sync(self, local_list, s3_list, bucket, prefix, acl, threads=16):
```
Comment (agamzn) [inline taskcat/s3_sync.py]: Super nitpicky: 
I think this function should follow either of the following.
- The variables should be named such that it self-documents (not the case right now)
- In-line comments should be added at each step so someone else could follow the logic (also not the case right now, in my opinion)

File: taskcat/s3_sync.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        pool.map(func, upload_to_s3)
+        pool.close()
+        pool.join()
+
+    def _s3_upload_file(self, paths, prefix, s3_client, acl):
+        local_filename, bucket, s3_path = paths
+        retry = 0
+        while retry < 5:
+            try:
+                s3_client.upload_file(local_filename, bucket, prefix + s3_path, ExtraArgs={'ACL': acl})
```
Comment (agamzn) [inline taskcat/s3_sync.py]: What's the behavior if `acl` is not provided?

Comment (jaymccon) [reply taskcat/s3_sync.py]: updated to default to "private"

File: taskcat/s3_sync.py (hunk 3 above)
Comment (jaymccon) [reply taskcat/s3_sync.py]: done

File: taskcat/s3_sync.py (hunk 1 above)
Comment (jaymccon) [reply taskcat/s3_sync.py]: done

File: taskcat/s3_sync.py (hunk 2 above)
Comment (jaymccon) [reply taskcat/s3_sync.py]: done

**Comment by jaymccon** [comment]
@agamzn thanks for the review, I've addressed the comments, can you please give another pass

**Comment by agamzn** [comment]
While all checks have passed and I approve, I think one more approval is a good idea. This is a pretty significant change to the codebase.

**Comment by jaymccon** [comment]
@avattathil @vsnyc @sshvans could one of you please give this pr a once over

**Comment by agamzn** [comment]
@avattathil 's commentary has been migrated to #154 for tracking.

**Comment by jaymccon** [comment]
@vsnyc totally agree on this. Seeing as the pr has already been merged, I'll address getting the methods more succinct in a future pr

**Pull request merged**